src/common/providers/identity.ts

```diff
@@ -109,3 +375,598 @@ export function userRecordConstructor(wireData: Object): UserRecord {
+      );
+    }
+  }
+  if (
+    eventType === 'beforeSignIn' &&
+    (authRequest as BeforeSignInResponse).sessionClaims
+  ) {
+    const invalidClaims = nonAllowListedClaims.filter((claim) =>
+      (authRequest as BeforeSignInResponse).sessionClaims.hasOwnProperty(claim)
+    );
+    if (invalidClaims.length > 0) {
+      throw new HttpsError(
+        'invalid-argument',
+        `The sessionClaims claims "${invalidClaims.join(
+          ','
+        )}" are reserved and cannot be specified.`
+      );
+    }
+    if (
+      JSON.stringify((authRequest as BeforeSignInResponse).sessionClaims)
+        .length > claimsMaxPayloadSize
+    ) {
+      throw new HttpsError(
+        'invalid-argument',
+        `The sessionClaims payload should not exceed ${claimsMaxPayloadSize} characters.`
+      );
+    }
+    const combinedClaims = {
+      ...authRequest.customClaims,
+      ...(authRequest as BeforeSignInResponse).sessionClaims,
+    };
+    if (JSON.stringify(combinedClaims).length > claimsMaxPayloadSize) {
+      throw new HttpsError(
+        'invalid-argument',
+        `The customClaims and sessionClaims payloads should not exceed ${claimsMaxPayloadSize} characters combined.`
+      );
+    }
+  }
+}
+
+/**
+ * @internal
+ * Helper function to generate the update mask for the identity platform changed values
+ */
+export function getUpdateMask(
+  authResponse?: BeforeCreateResponse | BeforeSignInResponse
+): string {
+  if (!authResponse) {
+    return '';
+  }
+  const updateMask: string[] = [];
+  for (const key in authResponse) {
+    if (key === 'customClaims' || key === 'sessionClaims') {
+      continue;
+    }
+    if (
+      authResponse.hasOwnProperty(key) &&
+      typeof authResponse[key] !== 'undefined'
+    ) {
+      updateMask.push(key);
+    }
+  }
+  return updateMask.join(',');
+}
+
+/** @internal */
+export function createHandler(
+  handler: (
+    user: UserRecord,
+    context: AuthEventContext
+  ) =>
+    | BeforeCreateResponse
+    | Promise<BeforeCreateResponse>
+    | BeforeSignInResponse
+    | Promise<BeforeSignInResponse>
+    | void
+    | Promise<void>,
+  eventType: string
+): (req: express.Request, resp: express.Response) => Promise<void> {
+  const wrappedHandler = wrapHandler(handler, eventType);
+  return (req: express.Request, res: express.Response) => {
+    return new Promise((resolve) => {
+      res.on('finish', resolve);
+      resolve(wrappedHandler(req, res));
+    });
+  };
+}
+
+/** @internal */
+function wrapHandler(
+  handler: (
+    user: UserRecord,
+    context: AuthEventContext
+  ) =>
+    | BeforeCreateResponse
+    | Promise<BeforeCreateResponse>
+    | BeforeSignInResponse
+    | Promise<BeforeSignInResponse>
+    | void
+    | Promise<void>,
+  eventType: string
+) {
+  return async (req: express.Request, res: express.Response): Promise<void> => {
+    try {
+      const projectId = process.env.GCLOUD_PROJECT;
+      const publicKeys = await fetchPublicKeys();
```

Comment:
IIRC these keys rotate very infrequently. Can we keep a cache and compare against a key (e.g. exp) to know whether we need to refresh the cache? We might also need some retry logic if the key rotated early. It would be nice to cut a request out of the amortized cost of a function when users are interactively waiting for completion

Reply:
Added a key cache by performing a similar technique that the admin SDK uses except without using a Class. In the `cache-control` header, we can parse for a `max-age` property in order to figure out when the keys needs to be refreshed. I added in a single retry to verification that will force refresh the keys.